File: temporal/worker-ts/src/workflows/humanApproval.ts
import {
  proxyActivities,
  defineSignal,
  defineQuery,
  setHandler,
  condition,
  log,
} from '@temporalio/workflow';
import type * as activities from '../activities/approvalActivities';

const { sendApprovalRequest, notifyApprovalResult } = proxyActivities<typeof activities>({
  startToCloseTimeout: '15s',
  retry: { maximumAttempts: 3 },
});

// Signal and Query definitions
export const approvalSignal = defineSignal<[ApprovalSignalInput]>('approval');
export const approvalStatusQuery = defineQuery<ApprovalStatus>('approvalStatus');

export interface ApprovalInput {
  requester: string;
  item: string;
  timeout: string; // e.g., "5m"
}

export interface ApprovalSignalInput {
  approved: boolean;
  approver: string;
  reason: string;
}

export interface ApprovalStatus {
  item: string;
  requester: string;
  status: 'pending' | 'approved' | 'rejected' | 'timed_out';
  approver?: string;
  reason?: string;
  waitingSince: string;
}

export interface ApprovalResult {
  item: string;
  requester: string;
  status: string;
  approver?: string;
  reason?: string;
}

/**
 * Human Approval Workflow — TypeScript Variant
 *
 * DIFFERENCE FROM GO: Adds a QUERY handler so you can check the approval
 * status at any time without looking at the UI. Use:
 *   temporal workflow query --workflow-id <id> --name approvalStatus
 */
export async function humanApprovalWorkflow(input: ApprovalInput): Promise<ApprovalResult> {
  log.info('humanApprovalWorkflow started', {
    requester: input.requester,
    item: input.item,
    timeout: input.timeout,
  });

  // Internal state
  let signalReceived = false;
  let signalData: ApprovalSignalInput | undefined;
  const waitingSince = new Date().toISOString();
  let currentStatus: ApprovalStatus = {
    item: input.item,
    requester: input.requester,
    status: 'pending',
    waitingSince,
  };

  // Register signal handler
  setHandler(approvalSignal, (signal: ApprovalSignalInput) => {
    log.info('Approval signal received', { approved: signal.approved, approver: signal.approver });
    signalReceived = true;
    signalData = signal;
    currentStatus = {
      ...currentStatus,
      status: signal.approved ? 'approved' : 'rejected',
      approver: signal.approver,
      reason: signal.reason,
    };
  });

  // Register query handler — this is the TS-specific addition
  setHandler(approvalStatusQuery, () => currentStatus);

  // Step 1: Send notification that approval is needed
  await sendApprovalRequest(input.requester, input.item);
  log.info('Approval request sent', { item: input.item });

  // Step 2: Wait for signal with timeout
  const timeoutMs = parseDuration(input.timeout);
  const received = await condition(() => signalReceived, timeoutMs);

  // Step 3: Handle result
  let result: ApprovalResult;

  if (!received) {
    log.info('Approval timed out', { item: input.item, timeout: input.timeout });
    currentStatus.status = 'timed_out';
    result = {
      item: input.item,
      requester: input.requester,
      status: 'timed_out',
      reason: `No response within ${input.timeout}`,
    };
  } else if (signalData?.approved) {
    result = {
      item: input.item,
      requester: input.requester,
      status: 'approved',
      approver: signalData.approver,
      reason: signalData.reason,
    };
  } else {
    result = {
      item: input.item,
      requester: input.requester,
      status: 'rejected',
      approver: signalData?.approver,
      reason: signalData?.reason,
    };
  }

  // Step 4: Notify outcome
  try {
    await notifyApprovalResult(result.status, input.item, input.requester);
  } catch (err) {
    log.warn('Failed to send result notification', { error: String(err) });
  }

  return result;
}

function parseDuration(s: string): number {
  const match = s.match(/^(\d+)(ms|s|m|h)$/);
  if (!match) return 10 * 60 * 1000; // default 10m
  const value = parseInt(match[1], 10);
  switch (match[2]) {
    case 'ms': return value;
    case 's': return value * 1000;
    case 'm': return value * 60 * 1000;
    case 'h': return value * 3600 * 1000;
    default: return 10 * 60 * 1000;
  }
}

File: temporal/worker-go/workflows/human_approval.go
package workflows

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"

	"temporal-worker-go/activities"
)

// ApprovalInput contains the request details.
type ApprovalInput struct {
	Requester string `json:"requester"`
	Item      string `json:"item"`
	Timeout   string `json:"timeout"` // e.g., "5m", "10m"
}

// ApprovalSignal is sent via the "approval" signal channel.
type ApprovalSignal struct {
	Approved bool   `json:"approved"`
	Approver string `json:"approver"`
	Reason   string `json:"reason"`
}

// ApprovalResult holds the final outcome.
type ApprovalResult struct {
	Item      string `json:"item"`
	Requester string `json:"requester"`
	Status    string `json:"status"` // "approved", "rejected", "timed_out"
	Approver  string `json:"approver,omitempty"`
	Reason    string `json:"reason,omitempty"`
}

// HumanApprovalWorkflow blocks on a signal to simulate manual intervention.
// Demonstrates: signals, human-in-the-loop, configurable timeouts.
//
// Signal name: "approval"
// Signal payload: ApprovalSignal JSON
func HumanApprovalWorkflow(ctx workflow.Context, input ApprovalInput) (*ApprovalResult, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("HumanApprovalWorkflow started",
		"requester", input.Requester,
		"item", input.Item,
		"timeout", input.Timeout,
	)

	activityOpts := workflow.ActivityOptions{
		StartToCloseTimeout: 15 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			MaximumAttempts: 3,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOpts)

	// Step 1: Send notification that approval is needed
	err := workflow.ExecuteActivity(ctx, activities.SendApprovalRequest, input.Requester, input.Item).Get(ctx, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to send approval request: %w", err)
	}
	logger.Info("Approval request sent", "item", input.Item)

	// Step 2: Wait for signal with timeout
	timeout, err := time.ParseDuration(input.Timeout)
	if err != nil {
		timeout = 10 * time.Minute // default
	}

	signalCh := workflow.GetSignalChannel(ctx, "approval")

	var signal ApprovalSignal
	var received bool

	// Create a selector to wait for either signal or timeout
	selector := workflow.NewSelector(ctx)

	selector.AddReceive(signalCh, func(ch workflow.ReceiveChannel, more bool) {
		ch.Receive(ctx, &signal)
		received = true
	})

	timerCtx, timerCancel := workflow.WithCancel(ctx)
	selector.AddFuture(workflow.NewTimer(timerCtx, timeout), func(f workflow.Future) {
		// Timer fired — timeout
		received = false
	})

	selector.Select(ctx)
	timerCancel() // Cancel timer if signal was received first

	// Step 3: Handle the result
	var result ApprovalResult
	result.Item = input.Item
	result.Requester = input.Requester

	if !received {
		// Timed out
		logger.Info("Approval timed out", "item", input.Item, "timeout", input.Timeout)
		result.Status = "timed_out"
		result.Reason = fmt.Sprintf("No response within %s", input.Timeout)
	} else if signal.Approved {
		logger.Info("Approval granted", "item", input.Item, "approver", signal.Approver)
		result.Status = "approved"
		result.Approver = signal.Approver
		result.Reason = signal.Reason
	} else {
		logger.Info("Approval rejected", "item", input.Item, "approver", signal.Approver)
		result.Status = "rejected"
		result.Approver = signal.Approver
		result.Reason = signal.Reason
	}

	// Step 4: Notify the outcome
	err = workflow.ExecuteActivity(ctx, activities.NotifyApprovalResult, result.Status, input.Item, input.Requester).Get(ctx, nil)
	if err != nil {
		logger.Warn("Failed to send result notification", "error", err)
		// Non-fatal — workflow still completes
	}

	return &result, nil
}
